test/unit/protocol/control_layer/ResendResponseNoResend.test.js

```diff
@@ -1,16 +1,23 @@
 import assert from 'assert'
 
-import ResendResponseNoResend from '../../../../src/protocol/control_layer/resend_response_no_resend/ResendResponseNoResend'
-import ResendResponseNoResendV1 from '../../../../src/protocol/control_layer/resend_response_no_resend/ResendResponseNoResendV1'
+import ResendResponseNoResend from '../../../../src/protocol/control_layer/resend_response/ResendResponseNoResend'
+import ControlMessage from '../../../../src/protocol/control_layer/ControlMessage'
+import ValidationError from '../../../../src/errors/ValidationError'
 
 describe('ResendResponseNoResend', () => {
+    describe('validation', () => {
+        it('throws on null requestId', () => {
+            assert.throws(() => new ResendResponseNoResend(ControlMessage.LATEST_VERSION, null, 'streamId', 0), ValidationError)
```

Comment:
Wonder if this test should be in the V2 specific tests, since if in V3 the requestId is no longer required, we'd still want the V2 version to always fail on missing requestId, right? Same question about all the tests using `LATEST_VERSION` i.e. does this set us up to accidentally remove test coverage for V2 when fixing the `LATEST_VERSION` tests for V3?

Reply:
Yeah, hmm, I see your point. I don't have an immediate elegant fix for that. Just to elaborate on how I wanted things to be:

The message object now intends to be version-agnostic, although the version (annoyingly) does affect validation, like you identified. The tests of the message classes should also aim to be version agnostic.

What I would have liked to do is _drop_ the version parameter from the constructor of all `ControlMessage`s. It's only needed for version-specific validation rules. For Message Layer (`StreamMessage`), this actually worked (at least for the time being), because all new fields have default values.

One solution would be to have separate _version-specific validators_ and tests for them. It would solve your point, but OTOH for the moment it seems a bit overkill, so I didn't go for it right now.

For now there are just some version-specific `if`s in the validation code in the message class constructor. It works but isn't very elegant. As the logic is in the message class, I guess the tests for it should be in the message class tests as well, if the rules change with future versions.

Actually, for `requestId` specifically, there are already version-specific validation tests in `ControlMessage.test.js`. From `V2` all `ControlMessage`s require a `requestId`. If it isn't required in `V3` anymore, then this test would change to reflect. Similarly for individual messages, I guess.

I don't have a clear course of action for this right now, and it's about the dangers of potential future changes, so I propose we ignore it for now and let's try to address it if any of those potential future changes realizes.